File: src/simulation/boid.js
var MAX_SPEED = window.SIMULATION_CONSTANTS.BOID_MAX_SPEED;
var MAX_FORCE = window.SIMULATION_CONSTANTS.BOID_MAX_FORCE;
var DESIRED_SEPARATION = window.SIMULATION_CONSTANTS.BOID_DESIRED_SEPARATION;
var NEIGHBOR_DISTANCE = window.SIMULATION_CONSTANTS.BOID_NEIGHBOR_DISTANCE;
var BORDER_OFFSET = window.SIMULATION_CONSTANTS.BOID_BORDER_OFFSET;
var EPSILON = window.SIMULATION_CONSTANTS.EPSILON;
var render_size = window.SIMULATION_CONSTANTS.BOID_RENDER_SIZE;

// Predator avoidance parameters
var PREDATOR_RANGE = window.SIMULATION_CONSTANTS.PREDATOR_RANGE;
var PREDATOR_TURN_FACTOR = window.SIMULATION_CONSTANTS.PREDATOR_TURN_FACTOR;

function Boid(x, y, simulation) {
	var randomAngle = Math.random() * 2 * Math.PI;
	this.velocity = new Vector(Math.cos(randomAngle), Math.sin(randomAngle));
	this.position = new Vector(x, y);
	this.acceleration = new Vector(0, 0);
	this.simulation = simulation;
	this.render_size = render_size;
}

Boid.prototype = {

	render: function () {
		var directionVector = this.velocity.normalize().multiplyBy(this.render_size);
		var inverseVector1 = new Vector(- directionVector.y, directionVector.x);
		var inverseVector2 = new Vector(directionVector.y, - directionVector.x);
		inverseVector1 = inverseVector1.divideBy(3);
		inverseVector2 = inverseVector2.divideBy(3);

		this.simulation.ctx.beginPath();
		this.simulation.ctx.moveTo(this.position.x, this.position.y);
		this.simulation.ctx.lineTo(this.position.x + inverseVector1.x, this.position.y + inverseVector1.y);
		this.simulation.ctx.lineTo(this.position.x + directionVector.x, this.position.y + directionVector.y);
		this.simulation.ctx.lineTo(this.position.x + inverseVector2.x, this.position.y + inverseVector2.y);
		this.simulation.ctx.lineTo(this.position.x, this.position.y);
		this.simulation.ctx.strokeStyle = 'rgba(0, 0, 0, 0.3)';
		this.simulation.ctx.stroke();
		this.simulation.ctx.fillStyle = 'rgba(0, 0, 0, 0.3)';
		this.simulation.ctx.fill();
	},

	getCohesionVector: function (boids) {
		var totalPosition = new Vector(0, 0);
		var neighborCount = 0;
		for (var i = 0; i < boids.length; i++) {
			var boid = boids[i];
			if (this == boid) {
				continue;
			}

			var distance = this.position.getDistance(boid.position) + EPSILON;
			if (distance <= NEIGHBOR_DISTANCE) {
				totalPosition = totalPosition.add(boid.position);
				neighborCount++;
			}
		}

		if (neighborCount > 0) {
			var averagePosition = totalPosition.divideBy(neighborCount);
			return this.seek(averagePosition);
		} else {
			return new Vector(0, 0);
		}
	},

	seek: function (targetPosition) {
		var desiredVector = targetPosition.subtract(this.position);
		desiredVector.iFastSetMagnitude(MAX_SPEED);
		var steeringVector = desiredVector.subtract(this.velocity);
		steeringVector.iFastLimit(MAX_FORCE);
		return steeringVector;
	},

	getSeparationVector: function (boids) {
		var steeringVector = new Vector(0, 0);
		var neighborCount = 0;

		for (var i = 0; i < boids.length; i++) {
			var boid = boids[i];
			if (this == boid) {
				continue;
			}

			var distance = this.position.getDistance(boid.position) + EPSILON;
			if (distance > 0 && distance < DESIRED_SEPARATION) {
				var deltaVector = this.position.subtract(boid.position);
				deltaVector.iNormalize();
				deltaVector.iDivideBy(distance);
				steeringVector.iAdd(deltaVector);
				neighborCount++;
			}
		}

		if (neighborCount > 0) {
			var averageSteeringVector = steeringVector.divideBy(neighborCount);
			averageSteeringVector.iFastSetMagnitude(MAX_SPEED);
			averageSteeringVector.iSubtract(this.velocity);
			averageSteeringVector.iFastLimit(MAX_FORCE);
			return averageSteeringVector;
		} else {
			return new Vector(0, 0);
		}
	},

	getAlignmentVector: function (boids) {
		var perceivedFlockVelocity = new Vector(0, 0);
		var neighborCount = 0;

		for (var i = 0; i < boids.length; i++) {
			var boid = boids[i];
			if (this == boid) {
				continue;
			}

			var distance = this.position.getDistance(boid.position) + EPSILON;
			if (distance > 0 && distance < NEIGHBOR_DISTANCE) {
				perceivedFlockVelocity.iAdd(boid.velocity);
				neighborCount++;
			}
		}

		if (neighborCount > 0) {
			var averageVelocity = perceivedFlockVelocity.divideBy(neighborCount);
			averageVelocity.iFastSetMagnitude(MAX_SPEED);
			var steeringVector = averageVelocity.subtract(this.velocity);
			steeringVector.iFastLimit(MAX_FORCE);
			return steeringVector;
		} else {
			return new Vector(0, 0);
		}
	},

	getPredatorAvoidanceVector: function(predator) {
		if (!predator) {
			return new Vector(0, 0);
		}
		
		var distance = this.position.getDistance(predator.position) + EPSILON;
		if (distance > 0 && distance < PREDATOR_RANGE) {
			var avoidanceVector = this.position.subtract(predator.position);
			avoidanceVector.iFastNormalize();
			
			var avoidanceStrength = (PREDATOR_RANGE - distance) / PREDATOR_RANGE;
			avoidanceVector.iMultiplyBy(avoidanceStrength * PREDATOR_TURN_FACTOR);
			avoidanceVector.iFastLimit(MAX_FORCE * 1.5);
			
			return avoidanceVector;
		}
		
		return new Vector(0, 0);
	},

	flock: function (boids) {
		var cohesionVector = this.getCohesionVector(boids);
		var separationVector = this.getSeparationVector(boids);
		var alignmentVector = this.getAlignmentVector(boids);

		separationVector.iMultiplyBy(this.simulation.separationMultiplier);
		cohesionVector.iMultiplyBy(this.simulation.cohesionMultiplier);
		alignmentVector.iMultiplyBy(this.simulation.alignmentMultiplier);

		this.acceleration.iAdd(cohesionVector);
		this.acceleration.iAdd(separationVector);
		this.acceleration.iAdd(alignmentVector);
		
		if (this.simulation.predator) {
			var predatorAvoidanceVector = this.getPredatorAvoidanceVector(this.simulation.predator);
			this.acceleration.iAdd(predatorAvoidanceVector);
		}
	},

	bound: function () {
		if (this.position.x > this.simulation.canvasWidth + BORDER_OFFSET) {
			this.position.x = -BORDER_OFFSET;
		}
		if (this.position.x < -BORDER_OFFSET) {
			this.position.x = this.simulation.canvasWidth + BORDER_OFFSET;
		}
		if (this.position.y > this.simulation.canvasHeight + BORDER_OFFSET) {
			this.position.y = -BORDER_OFFSET;
		}
		if (this.position.y < -BORDER_OFFSET) {
			this.position.y = this.simulation.canvasHeight + BORDER_OFFSET;
		}
	},

	update: function () {
		this.velocity.iAdd(this.acceleration);
		this.velocity.iFastLimit(MAX_SPEED);
		this.position.iAdd(this.velocity);
		this.bound();
		this.acceleration.iMultiplyBy(0);
	},

	run: function (boids) {
		this.flock(boids);
		this.update();
		this.render();
	}
}; 

File: python_simulation/boid.py
"""
Boid class - Exact port of JavaScript src/simulation/boid.js

This MUST match exactly with the JavaScript implementation to ensure
identical flocking behavior.
"""

import math
import random
from typing import List, Optional, TYPE_CHECKING

from .vector import Vector
from .constants import CONSTANTS

if TYPE_CHECKING:
    from .simulation import Simulation
    from .predator import Predator

class Boid:
    def __init__(self, x: float, y: float, simulation: 'Simulation'):
        # Initialize with random direction (matches JavaScript)
        random_angle = random.random() * 2 * math.pi
        self.velocity = Vector(math.cos(random_angle), math.sin(random_angle))
        self.position = Vector(x, y)
        self.acceleration = Vector(0, 0)
        self.simulation = simulation
        self.render_size = CONSTANTS.BOID_RENDER_SIZE
    
    def getCohesionVector(self, boids: List['Boid']) -> Vector:
        """Calculate cohesion force - move towards average position of neighbors"""
        total_position = Vector(0, 0)
        neighbor_count = 0
        
        for boid in boids:
            if self == boid:
                continue
            
            distance = self.position.getDistance(boid.position) + CONSTANTS.EPSILON
            if distance <= CONSTANTS.BOID_NEIGHBOR_DISTANCE:
                total_position = total_position.add(boid.position)
                neighbor_count += 1
        
        if neighbor_count > 0:
            average_position = total_position.divideBy(neighbor_count)
            return self.seek(average_position)
        else:
            return Vector(0, 0)
    
    def seek(self, target_position: Vector) -> Vector:
        """Seek towards target position"""
        desired_vector = target_position.subtract(self.position)
        desired_vector.iFastSetMagnitude(CONSTANTS.BOID_MAX_SPEED)
        steering_vector = desired_vector.subtract(self.velocity)
        steering_vector.iFastLimit(CONSTANTS.BOID_MAX_FORCE)
        return steering_vector
    
    def getSeparationVector(self, boids: List['Boid']) -> Vector:
        """Calculate separation force - avoid crowding neighbors"""
        steering_vector = Vector(0, 0)
        neighbor_count = 0
        
        for boid in boids:
            if self == boid:
                continue
            
            distance = self.position.getDistance(boid.position) + CONSTANTS.EPSILON
            if distance > 0 and distance < CONSTANTS.BOID_DESIRED_SEPARATION:
                delta_vector = self.position.subtract(boid.position)
                delta_vector.iNormalize()
                delta_vector.iDivideBy(distance)
                steering_vector.iAdd(delta_vector)
                neighbor_count += 1
        
        if neighbor_count > 0:
            average_steering_vector = steering_vector.divideBy(neighbor_count)
            average_steering_vector.iFastSetMagnitude(CONSTANTS.BOID_MAX_SPEED)
            average_steering_vector.iSubtract(self.velocity)
            average_steering_vector.iFastLimit(CONSTANTS.BOID_MAX_FORCE)
            return average_steering_vector
        else:
            return Vector(0, 0)
    
    def getAlignmentVector(self, boids: List['Boid']) -> Vector:
        """Calculate alignment force - steer towards average heading of neighbors"""
        perceived_flock_velocity = Vector(0, 0)
        neighbor_count = 0
        
        for boid in boids:
            if self == boid:
                continue
            
            distance = self.position.getDistance(boid.position) + CONSTANTS.EPSILON
            if distance > 0 and distance < CONSTANTS.BOID_NEIGHBOR_DISTANCE:
                perceived_flock_velocity.iAdd(boid.velocity)
                neighbor_count += 1
        
        if neighbor_count > 0:
            average_velocity = perceived_flock_velocity.divideBy(neighbor_count)
            average_velocity.iFastSetMagnitude(CONSTANTS.BOID_MAX_SPEED)
            steering_vector = average_velocity.subtract(self.velocity)
            steering_vector.iFastLimit(CONSTANTS.BOID_MAX_FORCE)
            return steering_vector
        else:
            return Vector(0, 0)
    
    def getPredatorAvoidanceVector(self, predator: Optional['Predator']) -> Vector:
        """Calculate predator avoidance force"""
        if not predator:
            return Vector(0, 0)
        
        distance = self.position.getDistance(predator.position) + CONSTANTS.EPSILON
        if distance > 0 and distance < CONSTANTS.PREDATOR_RANGE:
            avoidance_vector = self.position.subtract(predator.position)
            avoidance_vector.iFastNormalize()
            
            avoidance_strength = (CONSTANTS.PREDATOR_RANGE - distance) / CONSTANTS.PREDATOR_RANGE
            avoidance_vector.iMultiplyBy(avoidance_strength * CONSTANTS.PREDATOR_TURN_FACTOR)
            avoidance_vector.iFastLimit(CONSTANTS.BOID_MAX_FORCE * 1.5)
            
            return avoidance_vector
        
        return Vector(0, 0)
    
    def flock(self, boids: List['Boid']) -> None:
        """Apply flocking behavior"""
        cohesion_vector = self.getCohesionVector(boids)
        separation_vector = self.getSeparationVector(boids)
        alignment_vector = self.getAlignmentVector(boids)
        
        # Apply multipliers (from simulation)
        separation_vector.iMultiplyBy(self.simulation.separation_multiplier)
        cohesion_vector.iMultiplyBy(self.simulation.cohesion_multiplier)
        alignment_vector.iMultiplyBy(self.simulation.alignment_multiplier)
        
        self.acceleration.iAdd(cohesion_vector)
        self.acceleration.iAdd(separation_vector)
        self.acceleration.iAdd(alignment_vector)
        
        # Add predator avoidance if predator exists
        if self.simulation.predator:
            predator_avoidance_vector = self.getPredatorAvoidanceVector(self.simulation.predator)
            self.acceleration.iAdd(predator_avoidance_vector)
    
    def bound(self) -> None:
        """Handle boundary wrapping"""
        if self.position.x > self.simulation.canvas_width + CONSTANTS.BOID_BORDER_OFFSET:
            self.position.x = -CONSTANTS.BOID_BORDER_OFFSET
        if self.position.x < -CONSTANTS.BOID_BORDER_OFFSET:
            self.position.x = self.simulation.canvas_width + CONSTANTS.BOID_BORDER_OFFSET
        if self.position.y > self.simulation.canvas_height + CONSTANTS.BOID_BORDER_OFFSET:
            self.position.y = -CONSTANTS.BOID_BORDER_OFFSET
        if self.position.y < -CONSTANTS.BOID_BORDER_OFFSET:
            self.position.y = self.simulation.canvas_height + CONSTANTS.BOID_BORDER_OFFSET
    
    def update(self) -> None:
        """Update boid physics"""
        self.velocity.iAdd(self.acceleration)
        self.velocity.iFastLimit(CONSTANTS.BOID_MAX_SPEED)
        self.position.iAdd(self.velocity)
        self.bound()
        self.acceleration.iMultiplyBy(0)  # Reset acceleration
    
    def run(self, boids: List['Boid']) -> None:
        """Main update loop - flocking + update"""
        self.flock(boids)
        self.update()
        # Note: render() is not needed for training, only for visualization 